### PR TITLE
lightning: unlimited grpc MaxCallRecvMsgSize for TikvImporter

### LINKED ISSUE #56114: lightning report "received message larger than max (18552234 vs. 4194304)" during duplicate resolution
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

currently, we use default max-message-size which is 4M, see

https://github.com/pingcap/tidb/blob/bda79ca6f4a6a0371332a9fd8585a4ecce1916ab/pkg/lightning/backend/local/local.go#L144

### 2. What did you expect to see? (Required)
no error
### 3. What did you see instead (Required)
err
### 4. What is your TiDB version? (Required)
8.1.0

### PR DESCRIPTION
**Pull request by fishiu**
### What problem does this PR solve?

Issue Number: close #56114

Problem Summary: Currently, we comply with the default max-message-size=4MiB, but TiKVImporter may receive max 4k KV on each response, and each KV can be about 4KiB+, which exceeds the default limit easily.

### What changed and how does it work?

Set grpc options `MaxCallRecvMsgSize` to math.MaxInt32 (2GiB), i.e., unlimited in most scenarios. This number also aligns with the default setting for MaxCall**Send**MsgSize in gRPC.

MaxCallRecvMsgSize is only for resource control and there will not be any extra memory consumption. Please refer to the discussions below.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by lance6716** [comment]
Can you explain why choose 32 MB, is it enough?

Or can we choose the value after we get the error message like DM?

https://github.com/pingcap/tiflow/blob/76ebb371bb629cc8ca64c04694f985f927634acf/dm/ctl/common/util.go#L131-L139

**Comment by fishiu** [comment]
> Can you explain why choose 32 MB, is it enough?
> 
> Or can we choose the value after we get the error message like DM?
> 
> https://github.com/pingcap/tiflow/blob/76ebb371bb629cc8ca64c04694f985f927634acf/dm/ctl/common/util.go#L131-L139

32 MB is not an precise size, I chose it only refering to the size mentioned in the issue. Or does @D3Hunter have any idea about the upper bounds in real scenarios?

The way in DM is better. Bust I first need to locate places that use this client connection and wrap the recv process with adjust-and-retry logics.

**Comment by fishiu** [comment]
> Can you explain why choose 32 MB, is it enough?
> 
> Or can we choose the value after we get the error message like DM?
> 
> https://github.com/pingcap/tiflow/blob/76ebb371bb629cc8ca64c04694f985f927634acf/dm/ctl/common/util.go#L131-L139

I inspected the grpc source code regarding MaxCallRecvMsgSize and concluded that this option is only for limiting (memory/bandwidth) resource usage, i.e. setting MaxCallRecvMsgSize to large number will not directly consume large amount of memory.

Therefore, the strategy to capture the error message and then retry with adjusted options seems to be equivalent with directly setting it to intMax initially.

Additionally, the MaxCallRecvMsgSize is still an experimental option (for many years though) in grpc-go, so currently I do not think it is necessary to add it into lightning's config. Initializing it as maxInt would be simple and clear.

**Comment by D3Hunter** [comment]
> > Can you explain why choose 32 MB, is it enough?
> > Or can we choose the value after we get the error message like DM?
> > https://github.com/pingcap/tiflow/blob/76ebb371bb629cc8ca64c04694f985f927634acf/dm/ctl/common/util.go#L131-L139
> 
> I inspected the grpc source code regarding [MaxCallRecvMsgSize](https://github.com/grpc/grpc-go/blob/d66fc3a1efa1dfb33dfedf9760528f1ac2b923b6/rpc_util.go#L670C1-L677C36) and concluded that this option is only a numeric check to limit (memory/bandwidth) resource usage, i.e. setting MaxCallRecvMsgSize to large number will not directly consume large amount of memory. Actually in contrast, MaxCallSendMsgSize is by default maxInt32 in grpc.
> 
> Therefore, the strategy to capture the error message and then retry with adjusted options seems to be equivalent with directly setting it to intMax initially.
> 
> Additionally, the MaxCallRecvMsgSize is still an experimental option (for many years though) in grpc-go, so currently I do not think it is necessary to add it into lightning's config. Initializing it as maxInt would be simple and clear.

nice insight

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: pkg/lightning/backend/local/local.go [hunk 1]
```diff
@@ -165,6 +168,7 @@ func (f *importClientFactoryImpl) makeConn(ctx context.Context, storeID uint64)
 		addr = store.GetAddress()
 	}
 	opts = append(opts,
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(unlimitedRPCRecvMsgSize)),
```
Comment (D3Hunter) [inline pkg/lightning/backend/local/local.go:171]: have you test it manually?

Comment (fishiu) [reply pkg/lightning/backend/local/local.go:171]: I ran and passed the lightning_duplicate_detection integration test in local, which covered the modified code.

Comment (D3Hunter) [reply pkg/lightning/backend/local/local.go:171]: I mean to reproduce a scenario in the issue that have many duplicates and TiKV send a grpc message larger that some size, say 32MiB

Comment (fishiu) [reply pkg/lightning/backend/local/local.go:171]: No. But I have tried to reduce MaxCallRecvMsgSize e.g. 200KB which did trigger the expected error.

I will try to compose a new integration test that will send 4MiB+ KV back.

Comment (D3Hunter) [reply pkg/lightning/backend/local/local.go:171]: o, that should be enough, please also update this in the PR description

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**